config/samples/dataplane_v1beta1_openstackdataplane_nmstate.yaml

```diff
@@ -0,0 +1,258 @@
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlane
+metadata:
+  name: openstack-edpm
+spec:
+  deployStrategy:
+    deploy: false
+  nodes:
+    edpm-compute-0:
+      role: edpm-compute
+      hostName: edpm-compute-0
+      ansibleHost: 192.168.122.100
+      node:
+        ansibleVars: |
+          ctlplane_ip: 192.168.122.100
+          internal_api_ip: 172.17.0.100
+          storage_ip: 172.18.0.100
+          tenant_ip: 172.19.0.100
+          fqdn_internal_api: edpm-compute-0.example.com
+          # edpm_network_config - nmstate
+          edpm_network_config_tool: 'nmstate'
+          network_state:
+            dns-resolver:
+              config:
+                server:
+                - 192.168.122.80
+            interfaces:
+              - name: eth0
+                type: ethernet
+                state: up
+              - name: br-ex
+                type: ovs-interface
+                state: up
+                mtu: 1500
+                ipv4:
+                  enabled: true
+                  address:
+                    - ip: "192.168.122.100"
```

Comment:
I think we would need templates for network_state where ansible vars are jinja substituted like we do for os-net-config[1].  After the IPAM patches merge user would know these IPs, vlans etc beforehand.

[1]  https://github.com/openstack-k8s-operators/edpm-ansible/blob/main/roles/edpm_network_config/templates/single_nic_vlans/single_nic_vlans.j2

Reply:
Indeed this is an example. A lot of things could(should!) be improved, but it wasn't the purpose of this PR since the beginning.

BTW, It was my first thought having an approach similiar to actual os-net-config templates, but that means that we need to "wrap" the `network` role and that is something that @slagle would like to avoid.

We will think about it after the patches will be merged though. :)

Reply:
not sure I understand what you mean by "wrap the network role". We need the templates so that ansible vars for IPs etc could be substituted when using IPAM.

Reply:
Sorry, I'll try to explain my point of view better.

in `edpm_network_config` role we rely on `linux_system_roles.network` [role](https://github.com/linux-system-roles/network) to configure interfaces via `nmstate`. 

We should add tasks like [these](https://github.com/openstack-k8s-operators/edpm-ansible/blob/23b28886268f8725c0772305c4ca5e7ef2a1aa67/roles/edpm_network_config/tasks/os_net_config.yml#L23-L43) to render template files that should be passed to the `network` role as variable but when i asked James about that he told me to call the role directly and users can/should configure the variables it provides directly.

I could also misunderstood his words BTW, I'm not against templating but IMHO this isn't the right PR for that.

Reply:
I think we should handle this similar to how we're working with os-net-config j2 files in the latest proposals here:
https://github.com/openstack-k8s-operators/edpm-ansible/pull/243
https://github.com/openstack-k8s-operators/dataplane-operator/pull/284

These PR's will stop creating the `network` key in the inventory configmap and instead `copy` the yaml / jinja2 content directly to the nodes via the `edpm_network_config` role. We can either:

1. Re-use the same `spec.networkConfig.template` CR interface and handle it in Ansible;
2. Add a new key in the `networkConfig` struct specifically for nmstate configuration, but still render out j2 so that we can still populate it using data from IPAM. The same way we do with `os-net-config` today.

I think option one is easy enough, we can just have an additional variable to define it as a nmstate config vs os-net-config. But ultimately, nesting it under that `networkConfig` struct will be a cleaner user experience and provide more consistency. Provided we use Ansible's `copy` module the same way we do for `os-net-config`, then it will be fine with jinja2 and I think that should address the concern Rabi is raising in this last point.
https://github.com/openstack-k8s-operators/edpm-ansible/pull/243/files#diff-9add306dc693f43a54d6d0e77739b1fa2cefa165ddb6f2f55a0f65e6d7043bc7R29

Reply:
@bshephar the additional var is already there! :D `edpm_network_config_tool` variable can take the values `nmstate` or `os-net-config` (which is the default). That flag let the user choose between `linux-system-role.network` config or the "standard' os-net-config.

I'd like to nest the configuration under the `networkConfig` struct, but given that os-net-config developers have already introducted the support for nmstate, does it have sense to use a different kind of config under the same var?